### PR TITLE
Save config upon exiting evolver context

### PR DESCRIPTION
**Pull request by jamienoss**

### CODE REVIEW

File: evolver/device.py [hunk 1]
```diff
@@ -92,3 +92,12 @@ def loop_once(self):
             self.evaluate_controllers()
         if self.enable_commit:
             self.commit_proposals()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.config_model.save(app_settings.CONFIG_FILE)
+
```
Comment (jamienoss) [inline evolver/device.py:101]: We could also add here...
```suggestion
  self.enable_control = False
  self.enable_commit = False
```

File: evolver/app/main.py [hunk 2]
```diff
@@ -28,8 +28,11 @@ async def lifespan(app: FastAPI):
...
         app.state.evolver = Evolver.create()
-    asyncio.create_task(evolver_async_loop())
-    yield
+
+    with app.state.evolver:
+        asyncio.create_task(evolver_async_loop())
+        yield
+
     # Shutdown:
     ...
```
Comment (jamienoss) [inline evolver/app/main.py:37]: Note that such "save" could go here also/instead, however, the more encapsulated place for it to live is in ``Evolver``.

**Review by amitschang — COMMENTED** [review]
I'm okay with it, but since it doesn't really guarantee that any modified internal state is propogated to the file we can't rely on it and should do `update` any time we want to change the config in a reliable way - I'm afraid it might get confusing

**Comment by jamienoss** [comment]
>but since it doesn't really guarantee that any modified internal state is propogated to the file

It's not supposed to. But that also depends on what you mean by _guarantee_. This will _always_ write the config if the app exits (as long as it's not from a signal interrupt or python interp crash etc).

However, if you also mean a guarantee that modified state gets dumped to file when modified, the ``update`` endpoint doesn't do this either as state can be modified outside of the update, i.e., within control/experiment code where the user is free to do so.

I think what you might be expecting is periodic snapshots, however, this is something additional, and not something that this PR is supposed to address.

**Comment by jamienoss** [comment]
>should do update any time we want to change the config in a reliable way

Yep, that's exactly what the ``update`` endpoint is for, this is not supposed to reflect any functionality of mutating the config.

**Review by amitschang — APPROVED** [review]

**Pull request merged**